dask/highlevelgraph.py

```diff
@@ -500,21 +446,249 @@ def _repr_html_(self, layer_index="", highlevelgraph_key=""):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self.construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def get_output_keys(self) -> Set:
+        if hasattr(self, "_cached_output_keys"):
+            return self._cached_output_keys
+        else:
+            if self.output_blocks is None or not self.name:
+                output_keys = self.keys()  # Materializes the graph!
+            else:
+                output_keys = {(self.name, block) for block in self.output_blocks}
+            self._cached_output_keys = output_keys
+        return self._cached_output_keys
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def _keys_to_indices(self, keys):
+        """Convert keys to output chunk/partition indices
+
+        Intended for keys of the form: `(name, index)`.
+        """
+        parts = set()
+        for key in keys:
+            try:
+                _name, _part = key
+            except ValueError:
+                continue
+            if _name != self.name:
+                continue
+            parts.add(_part)
+        return parts
+
+    def cull(
+        self, keys: set, all_hlg_keys: Iterable
+    ) -> tuple[Layer, Mapping[Hashable, set]]:
+
+        # If the `output_blocks` property is defined, the materialized graph
+        # will only include the necessary tasks to produce these key indices.
+        # Therefore, in these cases, "culling" should only require us to
+        # re-initialize the layer with a new ``output_blocks`` parameter. For
+        # other cases, we must cast the layer to a ``MaterializedLayer``.
+
+        if self.output_blocks is None or self.name is None:
+            # If output_blocks is None, we must
+            # materialize the layer for culling
+            return MaterializedLayer(
+                dict(self),
+                annotations=self.annotations,
+            ).cull(keys, all_hlg_keys)
+
+        # Cull the layer by resetting output_blocks
+        output_blocks = self._keys_to_indices(keys)
+        if output_blocks != set(self.output_blocks):
+            new_state = self.layer_state.copy()
+            new_state["output_blocks"] = output_blocks
+            new_state["annotations"] = self.annotations
+            culled_layer = self.reconstructor()(**new_state)
+        else:
+            culled_layer = self
+        culled_deps = culled_layer.layer_dependencies(keys, output_blocks=output_blocks)
+        return culled_layer, culled_deps
+
+    def __dask_distributed_pack__(
+        self,
+        all_hlg_keys: Iterable[Hashable],
+        known_key_dependencies: Mapping[Hashable, set],
+        client,
+        client_keys: Iterable[Hashable],
+    ) -> Any:
+
```

Comment:
Why is this protocol still necessary?  Could we just lift the entire graph and send it to the scheduler?  What stops this?

Reply:
So, to make this concrete, how about the following diff.  

```diff
diff --git a/distributed/client.py b/distributed/client.py
index f68570f7..bc3afc75 100644
--- a/distributed/client.py
+++ b/distributed/client.py
@@ -2904,7 +2904,6 @@ class Client(SyncMethodMixin):
 
             # Pack the high level graph before sending it to the scheduler
             keyset = set(keys)
-            dsk = dsk.__dask_distributed_pack__(self, keyset, annotations)
 
             # Create futures before sending graph (helps avoid contention)
             futures = {key: Future(key, self, inform=False) for key in keyset}
@@ -2912,7 +2911,7 @@ class Client(SyncMethodMixin):
             self._send_to_scheduler(
                 {
                     "op": "update-graph-hlg",
-                    "hlg": dsk,
+                    "hlg": ToPickle(dsk),
                     "keys": list(map(stringify, keys)),
                     "priority": priority,
                     "submitting_task": getattr(thread_state, "key", None),
```

I suspect that this doesn't actually work today, but I'd be curious why not, and if it's not doable.